lms/djangoapps/courseware/access.py

```diff
@@ -129,6 +133,87 @@ def has_access(user, action, obj, course_key=None):
 
 
 # ================ Implementation helpers ================================
+def _can_access_descriptor_with_start_date(user, descriptor, course_key):  # pylint: disable=invalid-name
+    """
+    Checks if a user has access to a descriptor based on its start date.
+
+    If there is no start date specified, grant access.
+    Else, check if we're past the start date.
+
+    Note:
+        We do NOT check whether the user is staff or if the descriptor
+        is detached... it is assumed both of these are checked by the caller.
+
+    Arguments:
+        user (User): the user whose descriptor access we are checking.
+        descriptor (AType): the descriptor for which we are checking access.
+    where AType is any descriptor that has the attributes .location and
+        .days_early_for_beta
+    """
+    start_dates_disabled = settings.FEATURES['DISABLE_START_DATES']
+    if start_dates_disabled and not is_masquerading_as_student(user, course_key):
+        return True
+    else:
+        now = datetime.now(UTC())
+        effective_start = _adjust_start_date_for_beta_testers(
+            user,
+            descriptor,
+            course_key=course_key
+        )
+        return (
+            descriptor.start is None
+            or now > effective_start
+            or in_preview_mode()
+        )
+
+
+def _can_view_courseware_with_prerequisites(user, course):  # pylint: disable=invalid-name
+    """
+    Checks if a user has access to a course based on its prerequisites.
+
+    If the user is staff or anonymous, immediately grant access.
+    Else, return whether or not the prerequisite courses have been passed.
```

Comment:
Not sure I agree.  This method is specific to prerequisite courses, and hence the terminology need not be the generic one related to milestones.